draft/real-world-phoenix-lets-deploy-using-kubernetes.md

```diff
@@ -0,0 +1,643 @@
+  image: node:alpine
+  before_script:
+    - cd assets
+
+# Compile our elixir artifacts
+elixir_compile:
+  <<: *elixir_default
+  stage: init
+  script:
+    - mix deps.get --only test
+    - mix compile
+    - mix compile --warnings-as-errors
+  artifacts:
+    paths:
+      - mix.lock
+      - _build
+      - deps
+
+# compile javascript artifacts
+javascript_deps:
+  <<: *javascript_default
+  stage: init
+  script:
+    - npm install --progress=false
+  artifacts:
+    paths:
+      - assets/node_modules
+
+# build our container image
+build:
+  stage: build
+  image:
+    name: gcr.io/kaniko-project/executor:debug
+    entrypoint: [""]
+  script:
+    - echo "{\"auths\":{\"$CI_REGISTRY\":{\"username\":\"$CI_REGISTRY_USER\",\"password\":\"$CI_REGISTRY_PASSWORD\"}}}" > /kaniko/.docker/config.json
+    - /kaniko/executor --context $CI_PROJECT_DIR --dockerfile $CI_PROJECT_DIR/Dockerfile --destination $CONTAINER_RELEASE_IMAGE
+
+# and deploy our kubernetes cluster
+deploy:
+  stage: deploy
+  image:
+    name: lwolf/kubectl_deployer:latest
+  script:
+    - cat deploy.yml | envsubst | kubectl apply -f -
+    - cat service.yml | envsubst | kubectl apply -f -
+    - cat ingress.yml | envsubst | kubectl apply -f -
+  only:
+    refs:
+      - master
+```
+
+### Yaml deployment files: service, ingress and deploy
+The last thing we need is the kubernetes yaml files that are used in the ci
+pipeline:
+
+* service.yml
+The service pod running our application. This is basically our app running in a docker container.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: real-world-phoenix-service
+spec:
+  ports:
+  - name: web
+    port: 8001
+    protocol: TCP
+  selector:
+    app: real_world_phoenix
+```
+
+* ingress.yml
+The mapping of our domain name to our service
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: real-world-phoenix-ingress
+spec:
+  rules:
+  - host: $DOMAIN
+    http:
+      paths:
+      - path: "/"
+        backend:
+          serviceName: real-world-phoenix-service
+          servicePort: web
+```
+
+* deploy.yml
+Our deployment configuration
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: real-world-phoenix
+  labels:
+    app: real_world_phoenix
+spec:
+  revisionHistoryLimit: 5
+  replicas: 1
+  selector:
+    matchLabels:
+      app: real_world_phoenix
+  template:
+    metadata:
+      labels:
+        app: real_world_phoenix
+    spec:
+      imagePullSecrets:
+        - name: "gitlab.com"
+      containers:
+      - name: real-world-phoenix
+        image: registry.gitlab.com/$CI_PROJECT_NAMESPACE/$CI_PROJECT_NAME:$CI_COMMIT_SHORT_SHA
+        imagePullPolicy: Always
+        name: real-world-phoenix-deployment
+        resources: {}
+        ports:
+          - containerPort: 8001
+        env:
+          - name: SECRET_KEY_BASE
+            value: $SECRET_KEY_BASE
+          - name: DB_USER
+            value: $DB_USER
+          - name: DB_PASSWORD
```

Comment:
Passwords should come from secrets even though secrets do not provide security.

Reply:
? Is there something I should do differently here? Gitlab does mask these variables as well in case something gets logged.

Reply:
If you do `kubectl get pod <your-pod>` it would show your password in plain text. I would create a secret and ad a secretref with which you pull in all variables in the secret as env vars.

Reply:
ah ok, I'll look into that some more, thanks!